### ISSUE TITLE
Feature request: Recursive direct TSQR

### ISSUE THREAD

**Issue by lsorber** [dask/dask#2675]
(Apologies for any errors, I'm new to dask.)

According to the authors of the Direct TSQR paper [1] referenced in `dask.array.linalg.tsqr`:
> A central limitation to the Direct TSQR method is the necessity of gathering all R factors from the first step onto one processor in the second step.

A quick analysis of the algorithm's memory usage:
- Assume a matrix of R rows and k columns (R >> k).
- Write R as a vertical concatenation of N blocks of size (pk x k).
- Step 1 of TSQR computes N QR factorizations of size (pk x k) each. In a distributed setting with only one thread per machine, this would require O(pk^2) memory.
- Step 2 of TSQR computes a QR factorization of an (Nk x k) matrix, which requires O(Nk^2) memory.
- In summary, the memory required at any given time is O(max(N,p)k^2).

To minimize the memory usage, this suggests to use N = sqrt(R/k) blocks of size (Nk x k) each. For example, an (R=10e6 x k=1e4) matrix would be divided into N = 32 blocks of size (320e3 x 1e4). Each block would require about 24 GB of memory.

To resolve this issue, the authors proposed a recursive version of TSQR, which would allow Dask to work around memory constraints for the user. I realize it's a bit of work though, so as a stopgap, perhaps simply adding some additional pointers about how to choose the block size and the corresponding expected memory usage in the documentation would be useful.

[1] http://arxiv.org/abs/1301.1071

**Comment by mrocklin** [comment]
This would be great to have.  Is this something that you are interested in contributing?

cc @marianotepper

**Comment by lsorber** [comment]
@mrocklin I'd like to help out, but it would be somewhere next year at the earliest. If no one else is up for it by then, I could try my hand at it. But even the non-recursive implementation can scale to pretty large amounts of data if it's tuned properly, so adding some documentation to help out with that wouldn't be a bad idea either.

**Comment by mrocklin** [comment]
It looks like the `tsqr` docstring contains this text

    If blocks are of size ``(n, k)`` then this algorithm has memory use that
    scales as ``n**2 * k * nthreads``.

Is there other guidance that we should put into that docstring?  Should we write this content somewhere else other than the docstring, like a dedicated docpage?

**Comment by lsorber** [comment]
There are a few items to consider yes*:
- To optimize memory usage, the number of blocks should be configured as N = sqrt(R/k). This assumes one thread, with more threads per machine the blocks should be even smaller.
- To optimize computation time with P processing units, the blocks should be of size (Pk x k).
- Given the above two choices for partitioning the data, there is a trade-off to be made.

* Based on the pretty quick analysis in the OP, should verify these conclusions.

**Comment by mrocklin** [comment]
Any interest in submitting a PR that includes these notes within the tsqr docstring?

**Comment by lsorber** [comment]
Sure, if you don't mind that it could take a while before I can spare the time.

As a reminder for myself:
- [ ] Verify analysis of memory usage and computation time.
- [ ] Write additional `tsqr` documentation with pointers to optimize performance and submit a PR.

**Comment by mrocklin** [comment]
Sounds great.  Thank you @lsorber !

**Comment by marianotepper** [comment]
I would be glad to help with this. If @lsorber does not mind, I will try to
take a look at this in the next couple of weeks.

**Comment by mrocklin** [comment]
This sounds great.  It's nice to see you around again :)

**Comment by lsorber** [comment]
@marianotepper Sure, that's fine for me too. Do you mean improving the documentation or implementing the recursive version?

**Comment by marianotepper** [comment]
@mrocklin <https://github.com/mrocklin> I never meant to disappear, but
life happens... Good to be back!

@lsorber I meant improving the documentation. I'm not sure that I'll find
enough free time in the next few weeks for implementing new stuff.
I'll take a look, read the TSQR paper again, and see how much effort/time
it would take to implement the recursive version.

**Comment by convexset** [comment]
Ah... but memory issues remain unresolved. As with any single core QR decomposition, the memory requirements scale with the size of the original array.
